sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XXE/XmlReaderSettingsValidator.cs

```diff
@@ -0,0 +1,106 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2020 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.XXE
+{
+    internal class XmlReaderSettingsValidator
+    {
+        private readonly SemanticModel semanticModel;
+
+        public XmlReaderSettingsValidator(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        public bool IsUnsafe(InvocationExpressionSyntax invocation, ISymbol settings, NetFrameworkVersion version)
+        {
+            var unsafeDtdProcessing = false;
+            var unsafeResolver = version == NetFrameworkVersion.Probably35 || version == NetFrameworkVersion.Between4And451;
+
+            var objectCreation = GetObjectCreation(settings, invocation, semanticModel);
+            var objectCreationAssignments = objectCreation.GetInitializerExpressions().OfType<AssignmentExpressionSyntax>();
+
+            var assignments = invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>()
+                .DescendantNodes()
+                .OfType<AssignmentExpressionSyntax>()
+                .Where(assignment => assignment.Left is MemberAccessExpressionSyntax memberAccess &&
+                                     IsXmlReaderSettings(memberAccess.Expression, semanticModel) &&
+                                     settings.Equals(semanticModel.GetSymbolInfo(memberAccess.Expression).Symbol));
+
+            foreach (var assignment in objectCreationAssignments.Union(assignments))
+            {
+                if (assignment.Left.NameIs("ProhibitDtd") ||
```

Comment:
is it possible to move the assignment name checks before the `semanticModel.GetSymbolInfo` call?

Reply:
I've actually did this at first but the problem is that this check needs to be done only for assignments and not for the object creation so the code complexity increases a bit to handle the checks. 

In the end I've decided to do it like this since we do the call to GetSymbolInfo only for the assignments done on XmlReaderSettings so there is a very low chance to do unnecessary calls.

Reply:
so `semanticModel.GetTypeInfo` is not as expensive of `semanticModel.GetSymbolInfo` ?

Reply:
Yes, as far as I can tell GetTypeInfo is at least twice as fast:

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i5-5575R CPU 2.80GHz (Broadwell), 1 CPU, 4 logical and 4 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4121.0), X86 LegacyJIT
  DefaultJob : .NET Framework 4.8 (4.8.4121.0), X86 LegacyJIT


|                   Method |       Mean |    Error |   StdDev | Ratio | RatioSD |
|------------------------- |-----------:|---------:|---------:|------:|--------:|
|      GetTypeWithTypeInfo |   677.5 us | 13.30 us | 14.23 us |  1.00 |    0.00 |
| GetTypeWithGetSymbolInfo | 1,462.3 us | 29.00 us | 32.23 us |  2.16 |    0.06 |